helix-lsp-types: Replace `url::Url` type with String wrapper

the-mikedavis opened the pull request:
This replaces the use of the `url::Url` type with a newtype wrapper around a `String`. Changing URL type was the original motivation of vendoring/forking the `lsp_types` crate in https://github.com/helix-editor/helix/pull/11355. The `url` crate is based on the WHATWG spec for URLs which conflicts with the spec that LSP requires - RFC3986. This change uses the RFC3986 reserved characters for encoding paths which should fix some issues like #11888. The parsing and interpretation of the URL as a path is then covered by `Url::try_from` which we use ubiquitously since https://github.com/helix-editor/helix/pull/11486.

Fixes https://github.com/helix-editor/helix/issues/11888

the-mikedavis commented:
Exactly - this PR is working towards removing the url crate as a dependency and representing the URIs from language servers with a regular `String` instead.

RoloEdits commented:
Something I did notice in other work was the conversion of a url to a uri, `helix_core::uri::convert_url_to_uri`, was very expensive. It ends up calling `helix_stdx::path::normalize` every time.

![image](https://github.com/user-attachments/assets/c2164372-4c57-4c9c-af5d-71bf3c09ca95)


Not sure if the changes here will help the need for a conversion, mainly the need to normalize, or if this is entirely unavoidable, but thought I'd share at least. If there is a way to check beforehand that it is normalized, like some kind of fast bloom filter-like check, or modeling a just-in-time conversion for when paths are actually used, rather than an ahead-of-time conversion, this could be a big area of improvement.

This seems to happen most in the publish diagnostics hot path:

https://github.com/helix-editor/helix/blob/353176082e05696b292ccf65955333bd18deef97/helix-term/src/application.rs#L746-L753

the-mikedavis commented:
I'm not 100% sure we can avoid normalizing paths with this change. I imagine that a language server wouldn't send the same file as two different paths (for example `../dir/../a.txt` and `../a.txt`) especially since they're usually absolute paths, but maybe it's possible. I would say we can probably try not normalizing at first and add it back if needed.

the-mikedavis commented:
There are other schemes like zipfiles that could be similarly supported. But we would also probably want to do our own normalization for scratch buffers too.

the-mikedavis committed:
Swap `url::Url` for a lazy RFC3986 string wrapper

The `url::Url` crate covers the WHATWG URI which has different percent
encoding rules. LSP specifies that URIs are RFC3986. Strict language
servers would reject some cases like `[`/`]` characters in paths.

Dropping the `url` crate also drops a significant number of transitive
dependencies.

Co-authored-by: Blaž Hrastnik <blaz@mxxn.io>

RoloEdits commented:
Just for history, this is the profile I see now:
<img width="1890" height="825" alt="image" src="https://github.com/user-attachments/assets/aa67a05c-2dca-4e8d-9e96-cc1d4465c84b" />

The normalize calls completely disappeared. Whats left is just the small amount for `try_from` between Url and Uri. So big improvement here.

the-mikedavis commented:
Yep this change avoids normalization on the main path, it's mostly lazy now. ~It's nice to avoid the I/O / FS traversal work on the hot path~ (err whoops, I was thinking of canonicalization, not normalization). Maybe we can explore making the conversion to a Path a one-time cost too, to make it really cheap over the lifecycle of the editor.
